### PR TITLE
Change Detection Tutorial: factor in GRSM revisions

### PR DESCRIPTION
**Pull request by hkristen**
### Summary

Fixes premature early stopping that silently degraded test performance and adds reviewer-requested explanations for the pos_weight and backbone choices we did. Visualization now ranks test predictions by IoU (best/median/worst) instead of showing only cherry-picked samples.

### Rationale

The tutorial's early stopping (patience=10, min_epochs=1) could halt training before epoch 10 on the noisy 20-sample validation curve, producing much worse test metrics than reported. This also addresses GRSM reviewer feedback requesting justification for the Cityscapes backbone choice and the pos_weight=10 hyperparameter.

### Checklist

- [x] I have read the [Contributing docs](https://docs.torchgeo.org/en/stable/user/contributing.html)
- [x] I have read the [AI policy](https://github.com/torchgeo/governance/blob/main/AI-POLICY.md)

- [ ] 🟢 **No AI usage**: written by humans, for humans
- [x] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

**Comment by blaz-r** [comment]
I'm a bit out of the loop here, but given that I'm one of the authors on BTC, I'm interested in knowing what happened at GRSM? How and when was this reviewed? If you at any point would require more info about BTC, feel free to contact us.

**Comment by adamjstewart** [comment]
We are submitting a tutorial paper to GRSM (a journal, not a conference) showing off TorchGeo's new time-series support. This was one of the tutorials that we are in the process of improving.

**Comment by blaz-r** [comment]
Ah, cool. Thanks for letting me know. As for that cityscapes, as we also state in original paper, I think that the task is as importsnt if jot more than the domain of PT data. So segmentation pretrain on rgb data is wuite useful even if rgb data in finetune comes from rs domain. I guess you have handeled that already anyway.

**Comment by hkristen** [comment]
@calebrob6 @blaz-r could you review this? As @adamjstewart doesn't want to review AI-assisted PRs anymore. 

Thanks!

**Comment by calebrob6** [comment]
@adamjstewart did you submit work that depends on this PR?

**Comment by adamjstewart** [comment]
I'll review this, just give me some time.

File: docs/tutorials/change_detection.ipynb [hunk 1]
```diff
@@ -84,7 +102,7 @@
     "\n",
     "Let's download the dataset and look at some examples from OSCD100.\n",
     "\n",
-    "**About OSCD100**: This is a smaller, downsampled version of the full [OSCD](https://docs.torchgeo.org/en/stable/api/datasets/oscd.html#torchgeo.datasets.OSCD) (Onera Satellite Change Detection) dataset. It is designed for tutorials and quick experimentation. It contains 100 image pairs at 256×256 resolution with 60 training, 20 validation, and 20 test samples.\n",
+    "**About OSCD100**: This is a smaller, downsampled version of the full [OSCD](https://docs.torchgeo.org/en/stable/api/datasets/oscd.html#torchgeo.datasets.OSCD) (Onera Satellite Change Detection) dataset. It is designed for tutorials and quick experimentation. It contains 100 image pairs at 256\u00d7256 resolution with 60 training, 20 validation, and 20 test samples.\n",
```
Comment (calebrob6) [inline docs/tutorials/change_detection.ipynb:105]:
```suggestion
    "**About OSCD100**: This is a smaller, downsampled version of the full [OSCD](https://docs.torchgeo.org/en/stable/api/datasets/oscd.html#torchgeo.datasets.OSCD) (Onera Satellite Change Detection) dataset. It is designed for tutorials and quick experimentation. It contains 100 image pairs at 256 x 256 resolution with 60 training, 20 validation, and 20 test samples.\n",
```

File: docs/tutorials/change_detection.ipynb
Comment (calebrob6) [inline docs/tutorials/change_detection.ipynb:186]: is this comment necessary?

File: docs/tutorials/change_detection.ipynb [hunk 2]
```diff
@@ -180,7 +205,44 @@
...
+   "source": [
+    "We use [**BTC** (Be The Change)](https://docs.torchgeo.org/en/stable/api/models/btc.html), a change-detection-specific architecture that uses a [Swin Transformer](https://arxiv.org/abs/2103.14030) backbone pretrained on [Cityscapes](https://www.cityscapes-dataset.com/) followed by a [UPerNet](https://arxiv.org/abs/1807.10221) decoder. Unlike general segmentation models that simply concatenate both images into a single tensor, BTC processes them through separate encoder paths and fuses temporal features in a change-aware way.\n",
+    "\n",
+    "**Why a Cityscapes-pretrained backbone for satellite imagery?** Since we train on the three RGB bands, the input modality matches the RGB encoder exactly, so no spectral adaptation is needed. Cityscapes pretraining is dense-prediction pretraining on urban scenes, which transfers well to detecting urban change, and it is the pretraining used by the original BTC authors. If you instead train on all 13 Sentinel-2 bands, you should pick an encoder pretrained on multispectral satellite imagery, such as the Sentinel-2 Satlas weights shown in *To Go Further* below.\n",
```
Comment (calebrob6) [inline docs/tutorials/change_detection.ipynb:211]: is it possible to add a hyperlink to  *To Go Further*

Comment (calebrob6) [reply docs/tutorials/change_detection.ipynb:211]: Fun fact: the landcover.ai paper also got good results pre-training on cityscapes

File: docs/tutorials/change_detection.ipynb [hunk 3]
```diff
@@ -248,7 +312,7 @@
     "    fast_dev_run=fast_dev_run,\n",
     "    log_every_n_steps=1,\n",
     "    logger=logger,\n",
-    "    min_epochs=1,\n",
+    "    min_epochs=20,\n",
```
Comment (calebrob6) [inline docs/tutorials/change_detection.ipynb:315]: This needs to be mocked for testing in the same way fast_dev_run is mocked

File: docs/tutorials/change_detection.ipynb [hunk 4]
```diff
@@ -280,7 +346,7 @@
     "\n",
     "**In Google Colab**, run the cell below to launch TensorBoard inline.\n",
     "\n",
-    "**Locally**, run from your terminal `tensorboard --logdir /tmp/experiments` and navigate to https://localhost:6006"
+    "**Locally**, run from your terminal `tensorboard --logdir /tmp/experiments` and navigate to http://localhost:6006"
```
Comment (calebrob6) [inline docs/tutorials/change_detection.ipynb:349]: Why did this change? Does https work?

File: docs/tutorials/change_detection.ipynb [hunk 5]
```diff
@@ -324,7 +390,11 @@
    "cell_type": "markdown",
    "id": "25",
    "metadata": {},
-   "source": "## Visualize Predictions\n\nLet's visualize predictions on the test set. We create a fresh dataset (without the random crop augmentation the datamodule applies) so predictions align with the full 256×256 ground truth masks. We add the predicted mask to each sample and pass it to the dataset's built-in `plot()` method."
+   "source": [
+    "## Visualize Predictions\n",
+    "\n",
+    "Let's visualize predictions on the test set. We create a fresh dataset (without the random crop augmentation the datamodule applies) so predictions align with the full 256\u00d7256 ground truth masks. To get an honest picture of model performance rather than cherry-picked examples, we compute the IoU of every test sample and plot the best, median, and worst predictions. We add the predicted mask to each sample and pass it to the dataset's built-in `plot()` method.\n"
```
Comment (calebrob6) [inline docs/tutorials/change_detection.ipynb:396]:
```suggestion
    "Let's visualize predictions on the test set. We create a fresh dataset (without the random crop augmentation the datamodule applies) so predictions align with the full 256 x 256 ground truth masks. To get an honest picture of model performance rather than cherry-picked examples, we compute the IoU of every test sample and plot the best, median, and worst predictions. We add the predicted mask to each sample and pass it to the dataset's built-in `plot()` method.\n"
```

File: docs/tutorials/change_detection.ipynb (hunk 3 above)
Comment (adamjstewart) [reply docs/tutorials/change_detection.ipynb:315]: We stopped testing Case Studies by popular demand

**Comment by blaz-r** [comment]
I'll check this tomorrow morning (so like 12h) 👍

File: docs/tutorials/change_detection.ipynb (hunk 3 above)
Comment (calebrob6) [reply docs/tutorials/change_detection.ipynb:315]: So? It is still helpful to be able to test them when we want

File: docs/tutorials/change_detection.ipynb [hunk 6]
```diff
@@ -180,7 +205,44 @@
...
+    "\n",
+    "Key configuration choices:\n",
+    "- RGB bands (`in_channels=3`): Red, Green, Blue for a fast tutorial training experiment\n",
+    "- `weights=True`: Load Cityscapes-pretrained Swin-Tiny encoder\n",
+    "- `freeze_backbone=True`: Freeze the pretrained encoder and only fine-tune the decoder head, for faster convergence and less overfitting on this small dataset\n",
+    "- `pos_weight=10.0`: Upweights the change class to counter severe class imbalance, since most pixels show no change\n",
+    "- `lr=0.0001`: Use a small learning rate since we already have a pre-trained model\n",
+    "- `loss='bce'`: Binary cross-entropy loss\n",
```
Comment (blaz-r) [inline docs/tutorials/change_detection.ipynb:219]: BTC by default uses dice loss. Maybe that could also be used here? But this is not necessary by any means.

File: docs/tutorials/change_detection.ipynb [hunk 7]
```diff
@@ -257,7 +321,9 @@
    "cell_type": "markdown",
    "id": "19",
    "metadata": {},
-   "source": "**Note**: With a frozen backbone, training completes in ~50 seconds using ~2GB VRAM of your GPU."
+   "source": [
+    "**Note**: With a frozen backbone, training completes in ~50 seconds using ~2GB VRAM of your GPU."
```
Comment (blaz-r) [inline docs/tutorials/change_detection.ipynb:325]: Is this still true now with changed early stopping?

File: docs/tutorials/change_detection.ipynb [hunk 8]
```diff
@@ -337,26 +407,55 @@
...
     "    sample = viz_dataset[idx]\n",
     "    normed = datamodule.aug({'image': sample['image'].float().unsqueeze(0)})['image']\n",
     "    with torch.no_grad():\n",
     "        prob = task.predict_step({'image': normed.to(task.device)}, 0).squeeze().cpu()\n",
-    "    sample['prediction'] = (prob > 0.5).float().unsqueeze(0)\n",
-    "    fig = viz_dataset.plot(sample, suptitle=f'Test Sample {idx}')\n",
+    "    pred = prob > 0.5\n",
+    "    gt = sample['mask'].bool()\n",
+    "    union = (pred | gt).sum().item()\n",
+    "    iou = (pred & gt).sum().item() / union if union else 1.0\n",
```
Comment (blaz-r) [inline docs/tutorials/change_detection.ipynb:419]: I am not a fan of manual metric implementation, given how many papers do that and then (often unintentionally) get wrong results (mF1 vs F1 etc.). This is also discussed in the BTC paper itself 😄 where we also argue for the use of F1 as the primary metric for CD (since that is most common and makes everyone's life easier when comparing stuff). Early stopping also uses F1 in this tutorial.

I think the tutorials should promote good practices and just use torchmetrics F1 as the trainers do. I suggest [BinaryF1Score](https://lightning.ai/docs/torchmetrics/stable/classification/f1_score#binaryf1score) and then use its forward to get per-sample f1.

**Review by blaz-r — CHANGES_REQUESTED** [review]

File: docs/tutorials/change_detection.ipynb (hunk 8 above)
Comment (adamjstewart) [reply docs/tutorials/change_detection.ipynb:419]: Task computes these all automatically

Comment (blaz-r) [reply docs/tutorials/change_detection.ipynb:419]: Right, but this tutorial does this to pick samples to visualise, and I think it'd be good to use the torchmetric impl and F1 instead of iou. From a correctness standpoint, this is, of course, not an issue, and I ofc will not insist here if others don't see it as necessary, but I think it's good to use a standard implementation in case someone builds on top of this to prevent issues down the line.

Comment (adamjstewart) [reply docs/tutorials/change_detection.ipynb:419]: Task visualizes samples automatically. We don't need to visualize the best, just a few random ones. Simpler is better

Comment (blaz-r) [reply docs/tutorials/change_detection.ipynb:419]: All I'm saying is: use torchmetrics impl instead of handcrafted iou for the notebook.
